src/libPMacc/include/memory/IndexPool.hpp

```diff
@@ -0,0 +1,157 @@
+/* Copyright 2017 Heiko Burau
+ *
+ * This file is part of libPMacc.
+ *
+ * libPMacc is free software: you can redistribute it and/or modify
+ * it under the terms of either the GNU General Public License or
+ * the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * libPMacc is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License and the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * and the GNU Lesser General Public License along with libPMacc.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "pmacc_types.hpp"
+#include "memory/Array.hpp"
+#include <type_traits>
+
+namespace PMacc
+{
+namespace memory
+{
+
+    /* A memory pool of dynamic size containing indices
+     *
+     * \tparam T_Index type of index
+     * \tparam T_maxSize maximum number of indices
+     *
+     * Warning: This class is not thread-safe!
+     */
+    template<
+        typename T_Index,
+        size_t T_maxSize
+    >
+    struct IndexPool
+    {
+    private:
+
+        /* Reverse-iterator of the memory pool. The pool is iterated reversely
+         * to ensure removal of the current element while iterating.
+         */
+        struct ReverseIterator
+        {
+            T_Index* pointer;
+
+            HDINLINE
+            ReverseIterator( T_Index* const pointer ) : pointer( pointer )
+            {}
+
+            HDINLINE
+            void operator++()
+            {
+                this->pointer--;
+            }
+
+            HDINLINE
+            T_Index& operator*()
+            {
+                return *(this->pointer);
+            }
+
+            HDINLINE
+            bool operator!=( const ReverseIterator& other ) const
+            {
+                return this->pointer != other.pointer;
+            }
+        };
+
+        size_t m_size;
+        Array<
+            T_Index,
+            T_maxSize
+        > listIds;
+
+    public:
+
+        using Index = T_Index;
```

Comment:
the question is, if one actually needs that flexibility or just uses a meaningful type directly and writes a `.cpp` file to reduce compile time.

If it makes sense to keep a `T_Index`, please add at least a static assert to verify it is integer-like, e.g. via the trait [`std::is_integer`](http://en.cppreference.com/w/cpp/types/numeric_limits/is_integer) ( I don't even thing integral types would be useful)